### PR TITLE
type safe Conversion

### PR DESCRIPTION
**Pull request by filippor**
add method for typesafe Conversion

### CODE REVIEW

**Comment by cdietrich** [comment]
@filippor there is no ECA (https://www.eclipse.org/legal/ECA.php) signed for your signoff mail

**Comment by cdietrich** [comment]
i wonder if you could add some tests to org.eclipse.xtext.xbase.lib.tests
(and also wonder why there arent any yet)

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java [hunk 1]
```diff
@@ -61,139 +189,305 @@ public static Object doWrapArray(Object object) {
 
 		if (!arrayClass.isArray()) {
 			// Can't wrap something that isn't an array.
+//			throw new IllegalArgumentException("Can't wrap something that isn't an array.");
 			return object;
```
Comment (filippor) [inline org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: is it ok to return non wrapped object?
if not we change return type to Iterable<?>

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java [hunk 2]
```diff
@@ -42,16 +44,142 @@ private Conversions() {
 	}
 
 	/**
-	 * Wraps {@code object} in a list if and only if {@code object} is an array. Works for primitive and
-	 * object-component types.
+	 * Wraps {@code object} in a list if and only if {@code object} is an array.
```
Comment (szarnekow) [inline org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: Please try to avoid unrelated changes to the formatting. It makes digesting the PR harder than strictly necessary.

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java [hunk 3]
```diff
@@ -42,16 +44,142 @@ private Conversions() {
...
+	 * Works for primitive and object-component types.
+	 * 
+	 * @param object the object to be wrapped. May be <code>null</code>.
+	 * @return a list if the given object was an array. Otherwise the unmodified
+	 * given object. May return <code>null</code> if the object was
+	 * <code>null</code>.
+	 */
+	@Pure
+	public static WrappedBooleanArray doWrapArray(boolean[] object) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: I don't want to have the unfortunately publicly declared types WrappedXXXArray on the signatures of the methods here. What problem to you actually want to solve by this change?

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: This would change the contract of the method. I don't think that's possible.

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java [hunk 4]
```diff
@@ -61,139 +189,305 @@ public static Object doWrapArray(Object object) {
...
+	 * {@link Object} is used.
+	 * 
+	 * @param value the object to be unwrapped. May be <code>null</code>.
+	 * @return the previously wrapped array if the given value represents such.
+	 * Otherwise returns the value unmodified. May return <code>null</code> if
+	 * the value was <code>null</code>.
+	 */
+	@Pure
+	public static int[] unwrapArray(WrappedIntegerArray value) {
+		return value.internalToArray();
```
Comment (szarnekow) [inline org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: Missing null checks here and above

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java [hunk 5]
```diff
@@ -61,139 +189,305 @@ public static Object doWrapArray(Object object) {
...
+	 * {@link Object} is used.
+	 * 
+	 * @param value the object to be unwrapped. May be <code>null</code>.
+	 * @return the previously wrapped array if the given value represents such.
+	 * Otherwise returns the value unmodified. May return <code>null</code> if
+	 * the value was <code>null</code>.
+	 */
+
+	@Pure
+	public static <T> Object unwrapArray(Iterable<T> value) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: What's the benefit of the type parameter here?

File: org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java [hunk 6]
```diff
@@ -61,139 +189,305 @@ public static Object doWrapArray(Object object) {
...
+	 * <code>null</code>.
+	 * @return the previously wrapped array if the given value represents such.
+	 * Otherwise returns the value unmodified. May return <code>null</code> if
+	 * the value was <code>null</code>.
+	 * @throws ArrayStoreException if the expected runtime {@code componentType}
+	 * does not match the actual runtime component type.
 	 */
 	@Pure
-	public static Object unwrapArray(Object value, Class<?> componentType) {
+	public static <T,K extends T> Object unwrapArray(Iterable<K> value, Class<T> componentType) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: I don't think the signature is correct.
Assuming you have something like an array of strings and you store it in an iterable of objects.
Now K would point to object and T would need to be String. That does not fulfil the type constraints.
Also I don't see the parameter componentType to be used anymore, e.g. checkComponentType is missing.

Comment (filippor) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: so the orginal method that ignore componentType is wrong
`
if (value instanceof WrappedArray<?>) {
			Object result = ((WrappedArray<?>) value).internalToArray();
			return checkComponentType(result, componentType);
		}

`
and the check
`	private static Object checkComponentType(Object array, Class<?> expectedComponentType) {
		Class<?> actualComponentType = array.getClass().getComponentType();
		if (!expectedComponentType.isAssignableFrom(actualComponentType)) {
			throw new ArrayStoreException(
					String.format("The expected component type %s is not assignable from the actual type %s",
							expectedComponentType.getCanonicalName(), actualComponentType.getCanonicalName()));
		}
		return array;
	}`

Comment (filippor) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: this xtext test fail also with master code
 ```
@Test def fail(){
        var int[] array = #[0,1]
        var List<Integer> list = array
        var Integer[] array2 = list
        array2.forEach[v, index|
            assertEquals(index,v)
        ]        
    }
```

Comment (szarnekow) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: Does it fail with an assertion error or with a ClassCastException?

Comment (szarnekow) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: What do you mean with ‚original method that ignore checkComponentType‘?

Comment (filippor) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: the method on master if input value is WrappedXXXArray don't convert result to destination type
like it does for other Iterable

Comment (szarnekow) [reply org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java]: Sorry, I can’t follow. I see invocations of checkComponentType on master at the (to me) relevant locations.

**Comment by cdietrich** [comment]
@filippor @szarnekow what is the state of this pr?

File: org.eclipse.xtend.lib.tests/src/org/eclipse/xtend/lib/annotations/ArrayTest.xtend [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        list.set(0,1)
+        assertEquals(array.get(0),list.get(0))
+        var int[] array2 = list
+        assertTrue(array===array2)
+         
+    }
+    
+    @Ignore
+    @Test def fail(){
+        var int[] array = #[0,1]
```
[the remark this answers is not in the document]
Comment (filippor) [reply org.eclipse.xtend.lib.tests/src/org/eclipse/xtend/lib/annotations/ArrayTest.xtend]: it fail to convert WrappedIntArray to Integer[]

[the remark this answers is not in the document]
Comment (szarnekow) [reply org.eclipse.xtend.lib.tests/src/org/eclipse/xtend/lib/annotations/ArrayTest.xtend]: What's the purpose of this method if it is ignored?

**Comment by cdietrich** [comment]
@filippor any news?

**Comment by cdietrich** [comment]
ping @filippor

**Comment by cdietrich** [comment]
closed due missing feedback. feel free to reopen
